source/dataman/CLS/CLSDbUpgrade1Pt2.cpp

```diff
@@ -27,84 +28,132 @@ bool CLSDbUpgrade1Pt2::upgradeNecessary() const
+	QMap<int, QString> facilityTableThumbnails = queryAMFacilityTableThumbnails();
+	if (facilityTableThumbnails.count() == 1 && facilityTableThumbnails.values().at(0) == targetFacilityName_) {
+		databaseToUpgrade_->rollbackTransaction();
+		AMErrorMon::alert(this, CLSDbUpgrade1Pt2_ALREADY_UPDATED, QString("The thumbnail table has the correct data. No need to updgrade."));
+		return true;
+	}
 
-	// Stage 1: update the thumbnail table if required
-	//   - The AMDbObjectThumbnails_table doesn't have thumbnail for the given beamline name
-	QMap<int, QString> facilityTableThumbnails;
-	QSqlQuery thumbnailTableQuery = databaseToUpgrade_->select(thumbnailTableName, "id, title", QString("objectTableName=%1").arg("AMFacility_table") );
-	if (thumbnailTableQuery.exec()) {
-		while (thumbnailTableQuery.next()) {
-			facilityTableThumbnails.insert(thumbnailTableQuery.value(0).toInt(), thumbnailTableQuery.value(1).toString()); //id, title
+	// stage 1:
+	//         - update the thumbnail table
+	//           * remove the unnecessary thumbnail data for AMFacility table
+	//
+	foreach (int id, facilityTableThumbnails.keys()) {
+		if (facilityTableThumbnails.value(id) != targetFacilityName_) {
+			dbResult = databaseToUpgrade_->deleteRow(id, thumbnailTableName_);
+			if ( dbResult == 0 ){
+				AMErrorMon::alert(this, CLSDbUpgrade1Pt2_FAIL_TO_DELETE_FACILITY_THUMBNAIL, QString("Could not delete the thumbnail (%1) in %2").arg(id).arg(thumbnailTableName_));
+				databaseToUpgrade_->rollbackTransaction();
+				return false;
+			}
+
+			facilityTableThumbnails.remove(id);
 		}
 	}
-	thumbnailTableQuery.finish();
-
-	bool correctThumbnailTableData = (facilityTableThumbnails.count() == 1) && (facilityTableThumbnails.values().at(0) == targetFacilityName_);
-	if (!correctThumbnailTableData) {
-		foreach (int id, facilityTableThumbnails.keys()) {
-			if (facilityTableThumbnails.value(id) != targetFacilityName_) {
-				dbResult = databaseToUpgrade_->deleteRow(id, thumbnailTableName);
-				if ( dbResult == 0 ){
-					AMErrorMon::alert(this, CLSDbUpgrade1Pt2_COULD_NOT_DELETE_FACILITY, QString("Could not delete the thumbnail (%1) in %2").arg(id).arg(thumbnailTableName));
-					databaseToUpgrade_->rollbackTransaction();
-					return false;
-				}
-
-				facilityTableThumbnails.remove(id);
+
+	// stage 2:
+	//         - update the thumbnail table
+	//           *add new thumbnail data if there is no thumbnail for the target facility
+	//
+	if (facilityTableThumbnails.count() == 0) {
+
+		// query the facility information
+		int facilityId = -1;
+		AMFacility *facility = 0;
+
+		QSqlQuery queryFacility = databaseToUpgrade_->query();
+		queryFacility.prepare(QString("SELECT id, iconFileName FROM %1 where name='%2'").arg(facilityTableName_).arg(targetFacilityName_));
+
+		if (queryFacility.exec()) {
+
+			if (queryFacility.first()) {
+				facility = new AMFacility(targetFacilityName_, QString("CLS %1 Beamline").arg(targetFacilityName_), ":/clsIcon.png");
```

Comment:
Just use 

```c++
facilityImage = QImage(queryFacility.value(1).toString());
QBuffer imageBuffer;
bout.open(QIODevice::WriteOnly);
if(image.save(&imageBuffer, "PNG")) 
	thumbnail = imageBuffer.buffer();

.......
values << thumbnail;
```

Like I said, I would generally feel better if we don't use or instantiate any AMDbObjects.  There are aspects to that system that are linked together and can create unexpected behaviour.  I am quite certain that this won't have unintended consequences, but I don't want us to get in the habit of using them because database corruption is a very real possibility - especially for our delicate object/table relationships.